### DISCUSSION TITLE
Transparency and descender?

### DISCUSSION THREAD

**Discussion by Stanlyhalo** [Chlumsky/msdf-atlas-gen#59] [Q&A] [answered]
Hi, one very last thing before I go on my merry way. I'm using `ref.pixels` so when I do run the `savePng` function, it does save with transparency. But if I put the texture into memory by using this code:
```cpp
ImmediateAtlasGenerator<float, 4, mtsdfGenerator, BitmapAtlasStorage<byte, 4>> generator(width, height);
// ...
auto atlas = generator.atlasStorage();
msdfgen::BitmapRef<byte, 4> ref = atlas;
```
It does have black parts, and when I try to clip them in my shader code (clipping should work, it works with my other textures), it seems that those black parts that should be transparent as it is in the saved png, don't have the alpha channel set to 0.

On another note, I'm trying to add descender values for each glyph (so I can properly position certain glyphs like q, g, p, ...), and it works... just really scuffed and only works on certain fonts, on other it's totally messed up; obviously I'm not doing it right, so I was wondering if there's a way to get the value for each glyph, but the glyph's don't store the descender value anywhere. I'm sure I actually have to do math which is what I was doing before, but I'm not doing it right.

Edit1*: Just for more information, yes, I am setting my Vulkan format to `VK_FORMAT_R8G8B8A8_UNORM` to enable transparency. And bytes per pixel should be fine at 4. I know I can get descender from FontMetrics, but wouldn't each glyph have a unique descender?

Edit2*: Testing the FontMetrics descenderY, and it seems to be correct (at least for this font's testing), but of course, it effects all glyphs rn.

**Comment by Stanlyhalo** [comment]
Right now I'm using this:
![image](https://user-images.githubusercontent.com/36088749/205092446-2532558f-34d6-4a2d-9b9e-d2f845a60f47.png)

Although in the end, I need to divide it by some arbitrary value instead of based on the glyph's size (which I could change), so right now I divide by either 90/100... on some fonts like DIN Regular, it looks fine, than on other fonts it looks weird, and on other font it's too much or too little.

Everytime I get close to getting the descender to work properly, it breaks at different text sizes (text size is a value in my program, not in msdfgen). If I only use the size to correct the descender value, it works, but it's way bigger than expected. Do I just divide before the size multiplication by some arbitrary value? Because I've tried similar things but they break with different fonts.

**Comment by Stanlyhalo** [comment, marked as answer]
Okay, I think I got the descender working. I still have some things to work out but I'm still stuck on why ref.pixels either doesn't contain the alpha channel, or isn't being set by the generator properly.

Edit*: Answer is in replies.

**Reply by Chlumsky** [reply]
What makes you think there's something wrong with the alpha channel? Do you have an example of expected and actual contents?

**Reply by Stanlyhalo** [reply]
I actually am now thinking there might be something else that's wrong, I think I'm just encountering a weird engine bug of mine. Because my custom UI can render with the alpha channel just fine, but my 3d map cannot... so that solves that, now just onto adding ascenders (for dashes of course), and fixing look of the text (it looks kinda choppy even with the font scale set to something like 10.0). If you know how to fix any choppyness, that would be nice, I'll just fiddle with settings like in the atlas packer or maybe something else.

Edit*: Another reason it was hard to diagnose, is because a lot of methods of saving the png just ended up being the wrong format or something, so it couldn't save unless I just passed the ref.pixels to the savePng function, which had transparency in it, but not in my 3d map. Anyways, that is "solved" (bug known just not fixed lol)
